### PR TITLE
fix: better unsupported attribute support for aria-roledescription

### PR DESCRIPTION
**Pull request by straker**
Expand the `unsupported` property of an attribute by allowing certain elements to use the attribute.

Closes: #1216

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Has documentation updated, a DU ticket, or requires no documentation change
- [x] Includes new tests, or was unnecessary
- [x] Code is reviewed for security by: @WilcoFiers

### CODE REVIEW

File: doc/aria-supported.md [hunk 1]
```diff
@@ -14,8 +14,7 @@ For a detailed description about how accessibility support is decided, see [How
 
 ## Attributes
 
-| aria-attribute       | axe-core support |
-| -------------------- | ---------------- |
-| aria-describedat     | No               |
-| aria-details         | No               |
-| aria-roledescription | No               |
```
Comment (WilcoFiers) [inline doc/aria-supported.md]: We should do something to put the exceptions in here, something like "Everything except button, input[type=button], ...". Even better is we do something like "No [1]" and than under the table we put that list of exceptions as `1. Everything except ...`.

File: lib/checks/aria/unsupportedattr.js [hunk 2]
```diff
@@ -9,8 +9,33 @@ let unsupported = Array.from(node.attributes)
...
-	this.data(unsupported);
+if (!unsupportedAttrs.length) {
+	return false;
+}
+
+const nodeName = node.nodeName.toUpperCase();
+const lookupTable = axe.commons.aria.lookupTable;
+const role = axe.commons.aria.getRole(node);
+
+unsupportedAttrs = unsupportedAttrs.filter(attr => {
```
Comment (WilcoFiers) [inline lib/checks/aria/unsupportedattr.js]: I think this should be handled by `validateAttr`, instead of in the check.

File: build/tasks/aria-supported.js [hunk 3]
```diff
@@ -60,6 +61,33 @@ module.exports = function(grunt) {
 									!subject.hasOwnProperty(key)
 								) {
 									out.push([`${key}`, 'No']);
+								} else if (
```
Comment (jkodu) [inline build/tasks/aria-supported.js]: the complexity of this `reduce` fn has become hard to follow, worth splitting this into smaller fn's, with detailed comments.

File: lib/commons/aria/attributes.js [hunk 4]
```diff
@@ -37,13 +37,9 @@ aria.allowedAttr = function(role) {
  * @memberof axe.commons.aria
  * @instance
  * @param  {String} att The attribute name
- * @param  {Object} options Use `flagUnsupported: true` to report unsupported attributes
  * @return {Boolean}
  */
-aria.validateAttr = function(att, { flagUnsupported = false } = {}) {
+aria.validateAttr = function(att) {
```
Comment (jkodu) [inline lib/commons/aria/attributes.js]: I know you are only removing an argument here, could you name the function here, as it will be easier to debug in the long run.

```
aria.validateAttr = function validateAttr(att)....
```

File: build/tasks/aria-supported.js [hunk 5]
```diff
@@ -32,6 +32,46 @@ module.exports = function(grunt) {
...
+						const value = element.properties[prop];
+
+						// the 'type' property can be a string or an array
+						if (typeof value === 'string') {
+							return `\`<${element.nodeName} ${prop}="${value}">\``;
+						}
+
+						// output format for an array of types:
+						// <input type="button" | "checkbox">
+						let values = value.map(v => `"${v}"`).join(' | ');
```
Comment (jkodu) [inline build/tasks/aria-supported.js]: `let` -> `const`

File: build/tasks/aria-supported.js [hunk 6]
```diff
@@ -32,6 +32,46 @@ module.exports = function(grunt) {
...
 			const aQaria = new Set(axe.utils.uniqueArray(roleAriaKeys, ariaKeys));
+			let footnotes;
+
+			/**
+			 * Parse a list of unsupported exception elements and add a footnote
+			 * detailing which HTML elements are supported.
+			 * @param {Array<String|Object>} elements List of supported elements
+			 */
+			const generateSupportFootnote = elements => {
+				let supportedElements = elements.map(element => {
```
Comment (jkodu) [inline build/tasks/aria-supported.js]: `let` -> `const`

File: build/tasks/aria-supported.js [hunk 7]
```diff
@@ -32,6 +32,46 @@ module.exports = function(grunt) {
...
 			}, []);
 			const aQaria = new Set(axe.utils.uniqueArray(roleAriaKeys, ariaKeys));
+			let footnotes;
+
+			/**
+			 * Parse a list of unsupported exception elements and add a footnote
+			 * detailing which HTML elements are supported.
+			 * @param {Array<String|Object>} elements List of supported elements
+			 */
+			const generateSupportFootnote = elements => {
```
Comment (jkodu) [inline build/tasks/aria-supported.js]: Given that this is an extracted function, I do not see a reason for `footnotes` to be a top level variable.

I think we should scope `footnotes` to be within this function & return it, lying closer to the analogy of a pure function.

**Comment by jkodu** [comment]
@WilcoFiers 
I have updated this, so on your plate to review.

File: lib/checks/aria/unsupportedattr.js [hunk 8]
```diff
@@ -1,16 +1,33 @@
...
+	.filter(({ name }) => {
+		const attribute = lookupTable.attributes[name];
+		if (!axe.commons.aria.validateAttr(name) || !attribute.unsupported) {
+			return false;
+		}
+
+		const unsupported = attribute.unsupported;
+
+		if (typeof unsupported === 'boolean' || !unsupported.exceptions) {
+			return true;
```
Comment (WilcoFiers) [inline lib/checks/aria/unsupportedattr.js]: Nitpick, rather than checking if unsupported is falsey on line 8, I think this code would be a little easier to read if you check on line 14 if unsupported is not an object, and on line 14 do `return !!unsupported`.

File: lib/checks/aria/unsupportedattr.js [hunk 9]
```diff
@@ -1,16 +1,33 @@
...
+		}
+
+		// validate attributes and conditions (if any) from allowedElement to given node
+		let out = axe.commons.matches(node, unsupported.exceptions);
```
Comment (WilcoFiers) [inline lib/checks/aria/unsupportedattr.js]: `let out =` -> `const isException =`
